### ISSUE TITLE
`--optimize-runs` for Yul mode

### ISSUE THREAD

**Issue by Randall4** [ethereum/solidity#11708]
## Abstract

The Solidity compiler (solc) ignores the `--optimize-runs` parameter in Yul mode in favor of a list of Yul-specific optimizations. However, the default Yul optimizer list does not include runs-based optimizations that can improve gas efficiency, most notably for high numbers of expected runs.

## Motivation

It seems that some, if not all, of the optimizations done on Solidity contracts expecting a high number of runs are not included when compiling for Yul, so those using `solc` to compile Yul contracts aimed at high run counts are penalized.

Remix has a "Runs" field for Yul contracts and is able to apply these optimizations. Here is one example:

```
object "RunsTest1" {
    code {
        // Deploy the contract
        datacopy(0, dataoffset("Runtime"), datasize("Runtime"))
        return(0, datasize("Runtime"))
    }
    object "Runtime" {
        code {
            let funcSel := shl(224, 0xabc12345)
            mstore(0, funcSel)
        }
    }
}
```

When compiled with Runs = 1 in Remix, the output becomes
```
600b80600e600039806000f350fe63abc1234560e01b600052
PUSH1 0xB DUP1 PUSH1 0xE PUSH1 0x0 CODECOPY DUP1 PUSH1 0x0 RETURN POP INVALID PUSH4 0xABC12345 PUSH1 0xE0 SHL PUSH1 0x0 MSTORE
```

When compiled with Runs = 100000 in Remix, the "PUSH4 0xABC12345 PUSH1 224 SHL" gets compressed into a PUSH32 for gas savings:
```
602480600e600039806000f350fe7fabc1234500000000000000000000000000000000000000000000000000000000600052
PUSH1 0x24 DUP1 PUSH1 0xE PUSH1 0x0 CODECOPY DUP1 PUSH1 0x0 RETURN POP INVALID PUSH32 0xABC1234500000000000000000000000000000000000000000000000000000000 PUSH1 0x0 MSTORE
```

Using `solc` does not give this result (command `solc --yul --yul-dialect evm --optimize`)
```
600b80600d600039806000f3fe63abc1234560e01b600052
PUSH1 0xB DUP1 PUSH1 0xD PUSH1 0x0 CODECOPY DUP1 PUSH1 0x0 RETURN INVALID PUSH4 0xABC12345 PUSH1 0xE0 SHL PUSH1 0x0 MSTORE
```

`solc` works better than Remix for a number of other optimizations (for instance, Remix will output multiple `JUMPDEST`s in succession). Nonetheless, the Yul compiler would benefit greatly if some of these high-runs optimizations could be included.

## Specification

The `--optimize-runs` option should be made available in Yul mode. `--optimize --optimize-runs 100000` for the Yul code above should contain the PUSH compression optimization given above, for example.

## Backwards Compatibility

This feature will not introduce backward compatibility issues because the `--optimize-runs` option would be new for Yul mode.

**Comment by cameel** [comment]
I took a quick look at the CLI code and it's just that the `--optimize-runs` option is ignored in assembly mode (but the compiler does not report it as an invalid argument - we should really fix #11629).

So this should be very simple to fix - just pass the option down to `assemble()` in `CommandLineInterface.cpp`.

Fortunately this does not happen in Standard JSON mode and also this is the only optimizer option that is ignored without giving any feedback.

I'd classify it as a bug rather than a missing feature because this is a discrepancy with how Standard JSON works. It should have been changed back when the option was being added for Yul in Standard JSON.
